source/exponential_backoff_retry_strategy.c

```diff
@@ -0,0 +1,324 @@
+/*
+ * Copyright 2010-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <aws/io/retry_strategy.h>
+
+#include <aws/io/logging.h>
+
+#include <aws/common/clock.h>
+#include <aws/common/device_random.h>
+#include <aws/common/logging.h>
+#include <aws/common/mutex.h>
+#include <aws/common/task_scheduler.h>
+
+#include <inttypes.h>
+
+struct exponential_backoff_strategy {
+    struct aws_retry_strategy base;
+    struct aws_exponential_backoff_retry_config config;
+};
+
+struct exponential_backoff_retry_token {
+    struct aws_retry_token base;
+    struct aws_atomic_var current_retry_count;
+    struct aws_atomic_var last_backoff;
+    size_t max_retries;
+    uint64_t backofff_scale_factor;
+    enum aws_exponential_backoff_jitter_mode jitter_mode;
+    /* Let's not make this worst by constantly moving across threads if we can help it */
+    struct aws_event_loop *bound_loop;
+    uint64_t (*generate_random)(void);
+    struct aws_task retry_task;
+
+    struct {
+        struct aws_mutex mutex;
+        aws_retry_strategy_on_retry_token_acquired_fn *acquired_fn;
+        aws_retry_strategy_on_retry_ready_fn *retry_ready_fn;
+        void *user_data;
+    } thread_data;
+};
+
+static void s_exponential_retry_destroy(struct aws_retry_strategy *retry_strategy) {
+    if (retry_strategy) {
+        aws_mem_release(retry_strategy->allocator, retry_strategy);
+    }
+}
+
+static void s_exponential_retry_task(struct aws_task *task, void *arg, enum aws_task_status status) {
+    (void)task;
+
+    int error_code = AWS_ERROR_IO_OPERATION_CANCELLED;
```

Comment:
well retries live longer than a channel, a retry is likely on a new channel. The only cancelation we need to worry about here is the event-loop shutting down. I'm certainly not opposed to ref counting it though.

Reply:
actually, yeah, I think you're right... I'm going to keep the "release()" verb for the v-table though because it may or may not be a "destroy" operation.

Reply:
The API wouldn't have to change to alter this. It's not a one-way door. I vote for leaving it as is for now.